lib/core/src/display/shape/text/cursor.rs

```diff
@@ -100,43 +259,93 @@ lazy_static! {
+        let buffer_data      = cursors_vertices.flatten().collect_vec();
+        set_buffer_data(gl_context,self.buffer.as_ref().unwrap(),buffer_data.as_slice());
+        self.dirty = false;
+    }
+
     /// Removes all current cursors and replace them with single cursor without any selection.
     pub fn set_cursor(&mut self, position:CharPosition) {
-        self.cursors       = vec![Cursor::new(position)];
-        self.dirty_cursors = once(0).collect();
+        self.cursors = vec![Cursor::new(position)];
+        self.dirty   = true;
     }
 
     /// Add new cursor without selection.
     pub fn add_cursor(&mut self, position:CharPosition) {
-        let new_index = self.cursors.len();
         self.cursors.push(Cursor::new(position));
-        self.dirty_cursors.insert(new_index);
+        self.merge_overlapping_selections();
+        self.dirty = true;
     }
 
-    /// Update the cursors' buffer data.
-    pub fn update_buffer_data
-    (&mut self, gl_context:&Context, content:&mut TextComponentContent, fonts:&mut Fonts) {
-        let cursors          = self.cursors.iter();
-        let cursors_vertices = cursors.map(|cursor| Self::cursor_vertices(cursor,content,fonts));
-        let buffer_data      = cursors_vertices.flatten().collect_vec();
-        set_buffer_data(gl_context,&self.buffer,buffer_data.as_slice());
-        self.dirty_cursors.clear();
+    /// Do the navigation step of all cursors.
+    ///
+    /// If after this operation some of the cursors occupies the same position, or their selected
+    /// area overlap, they are irreversibly merged.
+    pub fn navigate_all_cursors(&mut self, navigaton:&mut CursorNavigation, step:&Step) {
+        self.cursors.iter_mut().for_each(|cursor| navigaton.move_cursor(cursor,&step));
+        self.merge_overlapping_selections();
+        self.dirty = true;
+    }
+
+    /// Number of vertices in cursors' buffer.
+    pub fn vertices_count(&self) -> usize {
+        self.cursors.len() * CURSOR_BASE_LAYOUT_SIZE
+    }
+
+    fn merge_overlapping_selections(&mut self) {
+        if !self.cursors.is_empty() {
+            self.cursors.sort_by_key(|c| c.position);
+            let mut last_range = self.cursors.first().unwrap().selection_range();
+            let mut i          = 1;
+            while i < self.cursors.len() {
+                let current_range = self.cursors[i].selection_range();
+                let merged        = Self::merged_selection_range(&last_range,&current_range);
+                match merged {
+                    Some(merged_range) => {
+                        self.cursors[i-1].extend_selection(&merged_range);
+                        self.cursors.remove(i);
+                        last_range = merged_range;
+                    },
+                    None               => {
+                        last_range = current_range;
+                        i += 1
+                    }
```

Comment:
In a case where multiple match branches are not line-by-line, I would not align `=>`. It looks strange. Btw, all toolsthat I know which align `=>` (like scalafmt) do it only when matches are line by line.